### PR TITLE
Refactor upstream tests to allow easier devfile implementation

### PR DESCRIPTION
**Pull request by ScrewTSW**
### What does this PR do?
Followup to discussions and changes done in this PR: https://github.com/eclipse/che/pull/15736

### What issues does this PR fix or reference?
#15829

### CODE REVIEW

File: tests/e2e/pageobjects/ide/Ide.ts [hunk 1]
```diff
@@ -218,32 +218,17 @@ export class Ide {
         await this.waitStatusBarContains(expectedTextInStatusBar, 20000);
     }
 
-    async closeAllNotifications() {
-        const notificationLocator: By = By.css('.theia-Notification');
-
+    async closeAllNotifications(timeout: number = TestConstants.TS_SELENIUM_DEFAULT_TIMEOUT) {
```
Comment (Ohrimenko1988) [inline tests/e2e/pageobjects/ide/Ide.ts:221]: WDYT about this variant?
```
    async closeAllNotifications(timeout: number = TestConstants.TS_SELENIUM_DEFAULT_TIMEOUT) {
        Logger.debug('Ide.closeAllNotifications');

        await this.clickStatusBarNotifications(timeout);
        await this.clickNotificationCenterCloseAllButton(timeout);
        await this.waitNotificationCenterClosedAll(timeout);
        await this.waitNotificationCenterDisappearance(timeout);
    }

    async clickStatusBarNotifications(timeout: number = TestConstants.TS_SELENIUM_DEFAULT_TIMEOUT) {
        Logger.debug('Ide.clickStatusBarNotifications');

        const statusBarNotificationsLocator: By = By.xpath(`//div[@id="theia-statusBar"]/div[@class="area right"]//div[contains(@title, 'Notification')]`);
        await this.driverHelper.waitAndClick(statusBarNotificationsLocator, timeout);
    }

    async clickNotificationCenterCloseAllButton(timeout: number = TestConstants.TS_SELENIUM_DEFAULT_TIMEOUT) {
        Logger.debug('Ide.clickNotificationCenterCloseAllButton');

        const notificationCenterActionCloseAllButtonLocator: By = By.xpath(`//div[contains(@class, 'theia-notifications-container theia-notification-center open')]//ul[@class="theia-notification-actions"]/li[@class="clear-all"]`);
        await this.driverHelper.waitAndClick(notificationCenterActionCloseAllButtonLocator, timeout);
    }

    async waitNotificationCenterClosedAll(timeout: number = TestConstants.TS_SELENIUM_DEFAULT_TIMEOUT) {
        Logger.debug('Ide.waitNotificationCenterClosedAll');

        const notificationCenterClosedLocator: By = By.xpath(`//div[contains(@class, 'theia-notifications-container theia-notification-center closed')]`);
        await this.driverHelper.waitPresence(notificationCenterClosedLocator, timeout);
    }

    async waitNotificationCenterDisappearance(timeout: number = TestConstants.TS_SELENIUM_DEFAULT_TIMEOUT) {
        Logger.debug('Ide.waitNotificationCenterDisappearance');

        const notificationCenterContainerNotificationsLocator: By = By.xpath(`//div[contains(@class, 'theia-notifications-container theia-notification-center')]//div[@class='theia-notification-list']//*`);
        await this.driverHelper.waitDisappearance(notificationCenterContainerNotificationsLocator, timeout);
    }
```
Besides when we should specify a place where intracted element is placed (NotificationCenter <--- CloseButton)
it is clear signal for creation new pageobject (NotificationCenter).

File: tests/e2e/utils/DriverHelper.ts [hunk 2]
```diff
@@ -521,6 +522,32 @@ export class DriverHelper {
         throw new error.TimeoutError(`Exceeded maximum mouse move attempts, for the '${elementLocator}' element`);
     }
 
+    public async hasChildren(elementLocator: By): Promise<boolean> {
```
Comment (Ohrimenko1988) [inline tests/e2e/utils/DriverHelper.ts]: What about this?
```
    public async hasChildren(elementXpathLocator: String): Promise<boolean> {
        return await this.isVisible(By.xpath(`${elementXpathLocator}//*`));
    }
```

File: tests/e2e/pageobjects/ide/Ide.ts (hunk 1 above)
Comment (ScrewTSW) [reply tests/e2e/pageobjects/ide/Ide.ts:221]: Hello Igor.
I'm happy to separate it into dedicated methods.

Regarding NotificationCenter, there already is an issue opened to create a pageobject for that and for notifications themselves https://github.com/eclipse/che/issues/15700 . I wouldn't include that into this PR and instead open it as a completely separate PR.

File: tests/e2e/utils/DriverHelper.ts (hunk 2 above)
Comment (ScrewTSW) [reply tests/e2e/utils/DriverHelper.ts]: :+1:

File: tests/e2e/pageobjects/ide/Ide.ts (hunk 1 above)
Comment (Ohrimenko1988) [reply tests/e2e/pageobjects/ide/Ide.ts:221]: Yes, a good idea, thank you for the answer.

**Comment by ScrewTSW** [comment]
@Ohrimenko1988 https://github.com/eclipse/che/pull/15736#discussion_r370638070
Radim recommended for me to refactor that method as a separate PR, that way we can only test that functionality and run E2E tests multiple times to ensure that the change is stable. I will create an issue for myself to not forget about this :)

**Comment by Ohrimenko1988** [comment]
@ScrewTSW Why don't to refactor it in this PR? It doesn't require a lot of time, besides I have already proposed a draft for you.

**Comment by Ohrimenko1988** [comment]
@ScrewTSW Anyway I see only one critical thing for me ("hasChildren"). All other only suggestions.

File: tests/e2e/testsLibrary/CodeExecutionTests.ts [hunk 3]
```diff
@@ -22,8 +25,25 @@ export function runTask(taskName: string, timeout: number) {
...
+        await topMenu.runTask(taskName);
+        await dialogWindow.waitDialogAndOpenLink(timeout, expectedDialogText);
+    });
+}
+
+export function runTaskWithDialogShellAndClose(taskName: string, expectedDialogText: string, timeout: number) {
+    test(`Run command '${taskName}' expecting dialog shell`, async () => {
+        await topMenu.runTask(taskName);
+        await dialogWindow.waitDialog(timeout, expectedDialogText);
+        await dialogWindow.clickToButton(dialogWindowCloseButtonText);
```
Comment (Ohrimenko1988) [inline tests/e2e/testsLibrary/CodeExecutionTests.ts]: Just for future:
```
        await dialogWindow.closeDialog()
```

**Review by Ohrimenko1988 — APPROVED** [review]

**Comment by ScrewTSW** [comment]
> @ScrewTSW Why don't to refactor it in this PR? It doesn't require a lot of time, besides I have already proposed a draft for you.

Increased complexity and recommended by @rhopp

File: tests/e2e/testsLibrary/CodeExecutionTests.ts (hunk 3 above)
Comment (ScrewTSW) [reply tests/e2e/testsLibrary/CodeExecutionTests.ts]: done

**Review by rhopp — APPROVED** [review]

**Comment by rhopp** [comment]
ci-build

**Pull request merged**